### PR TITLE
[release/7.0] Fix auth exception stopping QUIC connection listener

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/43723

## Description

If an HTTP/3 connection request from a client fails because of an auth issue, e.g. client rejects an invalid server certificate, then the connection listener accept loop stops. This prevents future connections.

## Customer Impact

Without this fix, a client can create a request that causes to Kestrel to stop accepting HTTP/3 connections.

## Regression?

- [ ] Yes
- [ ] No

Maybe. There might have been changes in System.Net.Quic to throw this error. [The XML docs on QuicListener.AcceptConnectionAsync](https://github.com/dotnet/runtime/blob/8582c5cbcf03d6c0d1d3f0d11e622b8b0168f50f/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs#L154-L158) have mixed signals about whether this exception should be thrown.

@dotnet/ncl could you confirm that this is expected behavior?

## Risk

- [ ] High
- [ ] Medium
- [x] Low

~This fixes this particular error, but I worry about if there are other exceptions that are thrown out of listener accept that would stop the connection accept loop. I'd like to get some advice on @dotnet/ncl on how best to use `QuicListener.AcceptConnectionAsync`.~

~We should keep listening if `AuthenticationException` is thrown. Is there a list of `QuicError` values for which we should continue listening? Or the inverse, what `QuicError` values should stop the accept loop?~

I think we have certainty from the NCL team on how best to use the QUIC connection listener.

## Verification

- [ ] Manual (required)
- [x] Automated

I'm unsure if it is possible to test this error in a browser manually. A browser HTTP/3 connection only works when the server offers a valid TLS cert; this failure only happens when the TLS cert is invalid.

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

@Pilchie

### CODE REVIEW

**Comment by rzikm** [comment]
> Also propagates exceptions from failed connection handshake, e.g. AuthenticationException, QuicException

I think the behavior is consistent with the documentation.

I realize it is not very user-friendly in the case of `AuthenticationException`, because the error is not caused by misconfiguring the QuicConnectionOptions, but swallowing this exception would make it hard for users to debug why they are not receiving any connections, Thoughts? @ManickaP?

**Comment by rzikm** [comment]
> I'd like to get some advice on https://github.com/orgs/dotnet/teams/ncl on how best to use QuicListener.AcceptConnectionAsync.
> 

> We should keep listening if AuthenticationException is thrown. Is there a list of QuicError values for which we should continue listening? Or the inverse, what QuicError values should stop the accept loop?

I believe only the `QuicError.OperationAborted` should stop the accept loop (because the listener was stopped), The rest of the errors/exceptions would be related to a failure of a specific connection which we did not accept.

**Comment by ManickaP** [comment]
The changed behavior was reaction to https://github.com/dotnet/runtime/issues/72160. Yes, the original issue was about misconfigured listener, but the decision was to expose all errors, not filter for specific ones and swallow the rest. Listener users can do that themselves, at least they have the option now.

> I believe only the QuicError.OperationAborted should stop the accept loop

Yes, I believe this as well.

**Comment by JamesNK** [comment]
While I think that `AcceptConnectionAsync` rethrowing the exception from the connection callback is well-intentioned, right now it means any exception thrown in `ConnectionOptionsCallback` can be rethrown. You could even create a callback to throw `new QuicException(QuicError.OperationAborted, ...)` and make someone think the listener has aborted (I don't know why someone would want to do that... just an suggestion of a problem of the current pattern).

A way to avoid this could be to wrap the exception thrown in `ConnectionOptionsCallback`, e.g. wrap it with `QuicException` and have a new QuicError of `ConnectionCallbackError`.

**Review by Tratcher — APPROVED** [review]

**Comment by Pilchie** [comment]
Approved for .NET 7 RC2.

**Comment by JamesNK** [comment]
@dotnet/aspnet-build Please merge.

**Comment by wtgodbe** [comment]
@JamesNK Are the test failures expected?

**Comment by JamesNK** [comment]
Oh. Tests. I'll take a look.

File: src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs [hunk 1]
```diff
@@ -145,38 +142,54 @@ public async ValueTask CreateListenerAsync()
...
+                // ObjectDisposedException is reported when an accept is started after the listener is unbind/disposed.
+                QuicLog.ConnectionListenerAborted(_log, ex);
+                return null;
+            }
+            catch (Exception ex)
+            {
+                // If the client rejects the connection because of an invalid cert then AcceptConnectionAsync throws.
+                // An error thrown inside ConnectionOptionsCallback can also throw from AcceptConnectionAsync.
+                // These are recoverable errors and we don't want to stop accepting connections.
+                QuicLog.ConnectionListenerAcceptConnectionFailed(_log, ex);
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:185]: > An error thrown inside ConnectionOptionsCallback can also throw from AcceptConnectionAsync.

Do exceptions originating from the  TlsHandshakeCallbackOptions.OnConnection 
get logged at a debug level when we're using the HttpsConnectionMiddleware?

Won't most exception types bubble out through AuthenticateAsServerAsync and get logged as an error by KestrelConnectionOfT? Ideally errors thrown by user code like the OnConnection callback should be logged at a higher level.

This is a fairly low-level API though, and it's best to get the fix in quickly, so I'll approve it either way.

File: src/Servers/Kestrel/Transport.Quic/test/QuicConnectionListenerTests.cs [hunk 2]
```diff
@@ -145,6 +187,87 @@ public async Task AcceptAsync_NoCertificateOrApplicationProtocol_Log()
...
+        // Assert
+        Assert.False(acceptTask.IsCompleted);
+
+        var clientConnection = await QuicConnection.ConnectAsync(options).DefaultTimeout();
+        var serverConnection = await acceptTask.DefaultTimeout();
+
+        //await Assert.ThrowsAsync<AuthenticationException>(() => QuicConnection.ConnectAsync(options).AsTask());
+        //Assert.Null(await acceptTask.AsTask().DefaultTimeout());
+
+        //Assert.Contains(LogMessages, m => m.EventId.Name == "ConnectionListenerAborted");
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Quic/test/QuicConnectionListenerTests.cs:268]: Uncomment? Did you mean to close the connection and verify there were no high-level logs?

**Review by halter73 — APPROVED** [review]

**Comment by JamesNK** [comment]
> @JamesNK Are the test failures expected?

I improved a flaky test. Ready to merge? 🙏

**Pull request merged**